### ISSUE TITLE
Reproducible segfault when switching tools with keybindings

### ISSUE THREAD

**Issue by antoine-levitt** [OpenBoard-org/OpenBoard#430]
Open openboard, begin drawing a line, maintain the mouse/pen down, change to eraser with ctrl-e while mouse/pen still down. This is openboard 1.6.1-a.1102 on linux. If people can't reproduce I can try posting a stacktrace.

(there are other bugs when switching tools, but this is the worst)

**Comment by letsfindaway** [comment]
I can reproduce that. When switching from pen to eraser, the segfault occurs at https://github.com/OpenBoard-org/OpenBoard/blob/85eb094ea9b876b559c8fd73e7fae1d7d9890903/src/domain/UBGraphicsScene.cpp#L1055, because `strokesGroup()` returns a `nullptr`.

I assume that switching tools while the mouse/pen is down is something which OpenBoard does not expect. What would you expect as intended behavior in this case - besides of course not causing a SEGFAULT? Should the first tool end at this point, just as with mouse up and the second tool start at the very same location?

**Comment by antoine-levitt** [comment]
Yes probably. Alternative is to emulate a mouse up and have the user make another mouse down to start, but just switching seems the most natural behavior.

The two other annoying bugs when changing tools with the keyboard are that it sometimes requires two switch to actually work, and that it leaves the overlay on. I can file issues if it's helpful. My use case is that my pen has buttons on it, which I've remapped to ctrl e and ctrl I for pen and eraser, which is very convenient

**Comment by antoine-levitt** [comment]
Generally without looking at the code it seems that the keyboard shortcuts bypass the events that are activated when the user switches using the buttons, and should instead call these functions

**Comment by letsfindaway** [comment]
The shortcuts are attached to the very same actions as the toolbar buttons, so they should invoke the same code. But the circumstances may vary. E.g. the pointer is on the board and not on the toolbar. The mouse is never down on the board when you switch tools with the toolbar etc... So I think it might be the circumstances which make the difference.

> The two other annoying bugs when changing tools with the keyboard are that it sometimes requires two switch to actually work, and that it leaves the overlay on.

It would help if you could find out what circumstances correspond to "sometimes". And what exactly do you mean with "overlay"?

**Comment by antoine-levitt** [comment]
By overlay I mean eg the circle around the pointer when using the eraser. So the two bugs are:

1 switch to eraser using the mouse, erase something. With the pointer on the board, do ctrl-i: the circle stays displayed

2 I can reproduce when pressing two shortcuts at once, but possibly also happens in other circumstances. Behavior is tricky to describe, but press ctrl-i, leave it pressed, then press ctrl-e. Both icons of pen and eraser are on, and then switching goes weird. If it doesn't work, do the reverse (press ctrl-e, leave it pressed, then press ctrl-i).

**Comment by letsfindaway** [comment]
> By overlay I mean eg the circle around the pointer when using the eraser.

I think this is also reported in #354. And all of that is probably related: tool usage is not properly terminated when switching with keyboard shortcuts while the tool is still active (mouse/pen down).

**Comment by kaamui** [comment]
Hello,

I pushed a fix for the segfault. I can't test at this time, so let me know if the behavior is more acceptable now without the segfault

Thanks !

**Comment by letsfindaway** [comment]
I just tested your fix: it avoids the segfault for me, but the handling of the switchover between the tools is still "strange":

I draw a line, then switch to eraser using Ctrl-E while mouse button still down, then erase something, then back to pen using Ctrl-I. This draws some line, probably from the last known pen position to my current position.

When I finally then release the mouse button, then the erased parts of the line reappear. So: the fix is better than nothing, as it avoids the crash, but still issues with changing tools while mouse button is down.

**Comment by letsfindaway** [comment]
This crash is also the same as reported in #329 from the original poster. So this bug is also fixed with the commits related to the current issue.

**Comment by PSeQXh** [comment]
Hello everyone, I just have a somehow remotly related question regarding keybindings: When I press ctrl+e in openboard v1.6.1 on Arch Linux, no tool changing happens. So are these mentioned keybindings OS specific? Thanks for what you do!

**Comment by letsfindaway** [comment]
Please don't attach your question to an existing "remotely related" issue but create a new one if you feel that your _specific_ issue isn't already covered elsewhere. It will also help if you give more details:

- From where did you install OpenBoard? Exactly specify the package used.
- What language translation are you using?
- You just mentioned Ctrl-E. Is it just Ctrl-E which doesn't work? Are there other shortcuts which are working? Which ones?
